slackbot/dispatcher.py

```diff
@@ -99,19 +100,26 @@ def loop(self):
             time.sleep(1)
 
     def _default_reply(self, msg):
-        try:
-            from slackbot_settings import default_reply
-        except ImportError:
-            default_reply = [
-                u'Bad command "{}", You can ask me one of the following questions:\n'.format(msg['text']),
-            ]
-            default_reply += [u'    • `{0}` {1}'.format(p.pattern, v.__doc__ or "")
-                              for p, v in six.iteritems(self._plugins.commands['respond_to'])]
-            # pylint: disable=redefined-variable-type
-            default_reply = u'\n'.join(default_reply)
-
-        m = Message(self._client, msg)
-        m.reply(default_reply)
+        if self._override_reply:
+            if six.callable(self._override_reply):
+                self._override_reply(self._client, msg)
+            else:
+                m = Message(self._client, msg)
+                m.reply(self._override_reply)
+        else:
+            if hasattr(settings, 'default_reply'):
+                default_reply = settings.default_reply
+            else:
+                default_reply = [
+                    u'Bad command "{}", You can ask me one of the following questions:\n'.format(msg['text']),
+                ]
+                default_reply += [u'    • `{0}` {1}'.format(p.pattern, v.__doc__ or "")
+                                  for p, v in six.iteritems(self._plugins.commands['respond_to'])]
+                # pylint: disable=redefined-variable-type
+                default_reply = u'\n'.join(default_reply)
+
+            m = Message(self._client, msg)
+            m.reply(default_reply)
```

Comment:
The above logic can be improved as follows:

1. find the per-instance default_reply
2. if no per-instance default_reply, use the global default_reply
3. if the global default_reply is not set as well, use the generated default reply 

Now we get the default_reply object, 

1. if it's a callable, call it with the `message` object as the argument
2. if it's a string, reply with it directly.